internal/proxy/powerscale_handler.go

```diff
@@ -0,0 +1,454 @@
+			s.log.Printf("decoding opa response: %+v", err)
+			writeError(w, "powerscale", "decoding opa request body", http.StatusInternalServerError)
+			return
+		}
+		log.Printf("OPA Response: %+v", opaResp)
+		if resp := opaResp.Result; !resp.Allow {
+			reason := strings.Join(opaResp.Result.Deny, ",")
+			s.log.Printf("request denied: %v", reason)
+			writeError(w, "powerscale", fmt.Sprintf("request denied: %v", reason), http.StatusBadRequest)
+			return
+		}
+
+		// At this point, the request has been approved.
+		// The driver does not send the volume request size so we set the Capacity to 0 to always approve the quota
+		qr := quota.Request{
+			SystemType:    "powerscale",
+			SystemID:      systemID,
+			StoragePoolID: isiPath,
+			Group:         group,
+			VolumeName:    pvName,
+			Capacity:      "0",
+		}
+
+		// Reset the original request
+		err = r.Body.Close()
+		if err != nil {
+			s.log.Printf("Failed to close original request body: %v", err)
+		}
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+		sw := &web.StatusWriter{
+			ResponseWriter: w,
+		}
+
+		s.log.Println("Proxying request...")
+		// Proxy the request to the backend powerflex.
+		r = r.WithContext(ctx)
+		next.ServeHTTP(sw, r)
+
+		// TODO(aaron): Determine if when the approved volume fails the volume is
+		// cleaned up (releasing capacity).
+		log.Printf("Resp: Code: %d", sw.Status)
+		switch sw.Status {
+		case http.StatusOK:
+			log.Println("Publish created")
+			ok, err := enf.PublishCreated(r.Context(), qr)
+			if err != nil {
+				log.Printf("publish failed: %+v", err)
+				return
+			}
+			log.Println("Result of publish:", ok)
+		default:
+			log.Println("Non 200 response, nothing to publish")
+		}
+	})
+}
+
+func (s *PowerScaleSystem) volumeDeleteHandler(next http.Handler, enf *quota.RedisEnforcement, opaHost string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx, span := trace.SpanFromContext(r.Context()).Tracer().Start(r.Context(), "volumeDeleteHandler")
+		defer span.End()
+
+		var systemID string
+		if v := r.Context().Value(web.SystemIDKey); v != nil {
+			var ok bool
+			if systemID, ok = v.(string); !ok {
+				writeError(w, "powerscale", http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+				return
+			}
+		}
+
+		isiPath := strings.TrimPrefix(filepath.Dir(r.URL.Path), "/namespace")
+		volName := filepath.Base(r.URL.Path)
+
+		b, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			writeError(w, "powerscale", "failed to read body", http.StatusInternalServerError)
+			return
+		}
+		defer r.Body.Close()
+
+		jwtValue := r.Context().Value(web.JWTKey)
+		jwtToken, ok := jwtValue.(token.Token)
+		if !ok {
+			writeError(w, "powerscale", "incorrect type for JWT token", http.StatusInternalServerError)
+			return
+		}
+
+		claims, err := jwtToken.Claims()
+		if err != nil {
+			writeError(w, "powerscale", "decoding token claims", http.StatusInternalServerError)
+			return
+		}
+
```

Comment:
Is there any validation done for a delete request?

Reply:
Yes, that should be done a few lines down where the `decision` package is used to query OPA.

Reply:
What is being evaluated? Or all delete request currently are allowed?

Reply:
The first OPA query validates the token. This should be the credential shield aspect. 

The quota query relates to checking if the tenant owns the volume being deleted.